File: digidis-go/day16/day16.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
	"time"
)

var (
	reg          = []int{0, 0, 0, 0}
	opcodes      = make(map[int]string)
	instructions = []string{"addr", "addi", "mulr", "muli", "banr", "bani", "borr", "bori", "setr", "seti", "gtir", "gtri", "gtrr", "eqir", "eqri", "eqrr"}
)

func main() {
	start := time.Now()
	defer func() {
		fmt.Printf("Done in %v\n", time.Since(start))
	}()

	do(true)
	do(false)

}

func do(part1 bool) {
	data, _ := ioutil.ReadFile("input-a.txt")
	rows := strings.Split(string(data), "\n")
	var before, code, after []int
	more3 := 0
	done := make(map[string]bool)

	// assuming one loop is enough for part2
	for _, r := range rows {

		if len(r) == 0 {
			m := 0
			var match []string
			for _, instr := range instructions {
				if !part1 && done[instr] {
					continue
				}
				copy(reg, before)
				run(instr, code)
				if eq(reg, after) {
					m++
					match = append(match, instr)
				}
			}
			if m == 1 {
				opcodes[code[0]] = match[0]
				done[match[0]] = true
			}
			if m >= 3 {
				more3++
			}
			before = nil
			after = nil

			continue
		}

		switch r[0] {
		case 'B':
			before = atois(r[9 : len(r)-1])
		case 'A':
			after = atois(r[9 : len(r)-1])
		default:
			code = atois(r)
		}

	}
	if part1 {
		fmt.Printf("More than 3 : %v\n", more3)
		return
	}

	reg = make([]int, 4)
	data, _ = ioutil.ReadFile("input-b.txt")
	rows = strings.Split(string(data), "\n")
	for _, r := range rows {
		code := atois(r)
		run(opcodes[code[0]], code)
	}
	fmt.Printf("Final registers %v\n", reg)
}

func run(op string, code []int) {
	switch op {
	case "addr":
		reg[code[3]] = reg[code[1]] + reg[code[2]]
	case "addi":
		reg[code[3]] = reg[code[1]] + code[2]
	case "mulr":
		reg[code[3]] = reg[code[1]] * reg[code[2]]
	case "muli":
		reg[code[3]] = reg[code[1]] * code[2]
	case "banr":
		reg[code[3]] = reg[code[1]] & reg[code[2]]
	case "bani":
		reg[code[3]] = reg[code[1]] & code[2]
	case "borr":
		reg[code[3]] = reg[code[1]] | reg[code[2]]
	case "bori":
		reg[code[3]] = reg[code[1]] | code[2]
	case "setr":
		reg[code[3]] = reg[code[1]]
	case "seti":
		reg[code[3]] = code[1]
	case "gtir":
		if code[1] > reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case "gtri":
		if reg[code[1]] > code[2] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case "gtrr":
		if reg[code[1]] > reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case "eqir":
		if code[1] == reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case "eqri":
		if reg[code[1]] == code[2] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case "eqrr":
		if reg[code[1]] == reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	}
}

func atoi(s string) int {
	i, _ := strconv.Atoi(s)
	return i
}

func atois(s string) (r []int) {
	s = strings.Replace(s, ",", "", -1)
	f := strings.Fields(s)
	for _, k := range f {
		r = append(r, atoi(k))
	}
	return
}

func eq(a []int, b []int) bool {
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

File: dlofstrom-python/day16.py
import sys
import re

# Read input
input = sys.stdin.read().strip()
input = input.split('\n\n')
# Samples come in sets of three
samples = [s.split('\n') for s in input[:-2]]
samples = [[list(map(int, re.findall('-?\d+',l))) for l in s] for s in samples]
# Example program
test_program = input[-1].split('\n')
test_program = [list(map(int, re.findall('-?\d+',l))) for l in test_program]

# Operations
def addr(A,B,C,registers):
    registers[C] = registers[A] + registers[B]
    return registers
    
def addi(A,B,C,registers):
    registers[C] = registers[A] + B
    return registers

def mulr(A,B,C,registers):
    registers[C] = registers[A] * registers[B]
    return registers

def muli(A,B,C,registers):
    registers[C] = registers[A] * B
    return registers

def banr(A,B,C,registers):
    registers[C] = registers[A] & registers[B]
    return registers

def bani(A,B,C,registers):
    registers[C] = registers[A] & B
    return registers

def borr(A,B,C,registers):
    registers[C] = registers[A] | registers[B]
    return registers

def bori(A,B,C,registers):
    registers[C] = registers[A] | B
    return registers

def setr(A,B,C,registers):
    registers[C] = registers[A]
    return registers

def seti(A,B,C,registers):
    registers[C] = A
    return registers

def gtir(A,B,C,registers):
    if A > registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def gtri(A,B,C,registers):
    if registers[A] > B:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def gtrr(A,B,C,registers):
    if registers[A] > registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def eqir(A,B,C,registers):
    if A == registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def eqri(A,B,C,registers):
    if registers[A] == B:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def eqrr(A,B,C,registers):
    if registers[A] == registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers


oplist = [addr, addi, mulr, muli, banr, bani, borr, bori, setr, seti, gtir, gtri, gtrr, eqir, eqri, eqrr]

# Initially every opcode can be any instruction
opcodes = {opcode:set(oplist) for opcode in set(o for o,a,b,c in [h[1] for h in samples])}
three_or_more = 0
for before,instruction,after in samples:
    opcode, a, b, c = instruction
    # Save possible instructions for opcode
    possible_operations = set(operation for operation in oplist if operation(a,b,c,list(before)) == after)
    # Note if 3 or more possible
    if len(possible_operations) >= 3:
        three_or_more += 1
    # Update possible instructions for opcode
    opcodes[opcode].intersection_update(possible_operations)
print('Part 1:', three_or_more)

# Those who only have one possible instruction can be eliminated from all others
while max(opcodes, key=lambda opcode: len(opcodes[opcode])) > 1:
    # Unique opcode instructions (union of sets)
    unique_operations = set.union(*[operations for operations in opcodes.values() if len(operations) == 1])
    # Not yet decided opcode instructions (list of sets)
    multiple_operations = [operations for operations in opcodes.values() if len(operations) > 1]
    for operations in multiple_operations:
        operations.difference_update(unique_operations)

# Make opcodes functions instead of sets
opcodes = {opcode:operations.pop() for opcode,operations in opcodes.items()}

# Run test program
registers = [0,0,0,0]
for opcode,a,b,c in test_program:
    opcodes[opcode](a,b,c,registers)
print('Part 2:', registers[0])
